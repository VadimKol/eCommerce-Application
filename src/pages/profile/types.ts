import type { Address } from '@commercetools/platform-sdk';

export interface CustomerProfile {
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  email: string;
  password: string;
  defaultShippingAddressId: string;
  defaultBillingAddressId: string;
}

export interface AddressOption {
  value: string;
  label: string;
}

export interface AddressCustom extends Address {
  id: string;
}
