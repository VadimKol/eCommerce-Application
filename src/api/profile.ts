import type { ClientResponse, Customer, MyCustomerChangePassword, MyCustomerUpdate } from '@commercetools/platform-sdk';

import { apiRoot } from './build-client';

export function profile(): Promise<ClientResponse<Customer>> {
  return apiRoot.me().get().execute();
}

export function changePassword(myCustomerChangePassword: MyCustomerChangePassword): Promise<ClientResponse<Customer>> {
  return apiRoot.me().password().post({ body: myCustomerChangePassword }).execute();
}

export function crudAddress(MyCustomerUpdate: MyCustomerUpdate): Promise<ClientResponse<Customer>> {
  return apiRoot.me().post({ body: MyCustomerUpdate }).execute();
}
