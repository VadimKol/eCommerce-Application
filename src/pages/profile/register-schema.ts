import { z } from 'zod';

import { calculateAge } from '@/common/utils';

export const registerSchema = z.object({
  email: z.string().email('Email address must be properly formatted (e.g., user@example.com)'),
  name: z.string().regex(/^[A-Za-z]+$/, 'Must contain at least one character and no special characters or numbers'),
  surname: z.string().regex(/^[A-Za-z]+$/, 'Must contain at least one character and no special characters or numbers'),
  age: z
    .string()
    .refine((date) => !Number.isNaN(Date.parse(date)), 'Invalid date format')
    .refine((date) => calculateAge(date), 'You must be at least 13 years old'),
});

export type RegisterSchema = z.infer<typeof registerSchema>;
