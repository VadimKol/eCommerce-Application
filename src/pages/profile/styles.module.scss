.main {
  padding: 20px;
}

.container {
  display: flex;
  flex-direction: column;
  gap: 10px;
  min-width: 850px;
  max-width: 100%;
  margin: auto;
  padding: 28px 40px 30px;
  text-align: center;
  color: var(--primary-color);
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(45deg, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;
}

.form {
  display: flex;
  flex-flow: row nowrap;
  gap: 15px;
  max-width: 600px;
  margin: 20px auto;
}

.profileBlock {
  display: flex;
  flex-flow: row nowrap;
  gap: 20px;
}

.detailPerson {
  width: 100%;
}

.button {
  width: 80px;
  height: 40px;
  margin: 30px auto 0;
  font-size: 1em;
  color: var(--text-color);
  background: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 6px;
  cursor: pointer;
  transition: all 500ms;
}

.button:hover {
  box-shadow: -5px 5px 5px var(--secondary-color);
  transition: all 500ms;
  transform: translateY(-2px);
}

.chooseInfoBlock {
  display: flex;
  flex-flow: column nowrap;
  gap: 20px;
  justify-content: start;
  align-items: center;
  width: 25%;
}

.chooseInfoItem {
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 200px;
  height: 100px;
  background: var(--background-color);
  border: 1px solid var(--secondary-color);
  border-radius: 10px;
  cursor: pointer;
}

.detailInfoBlock {
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  width: 75%;
}

.mainFormProfile {
  display: flex;
  flex-flow: column nowrap;
  gap: 20px;
  justify-content: center;
  align-items: start;
  margin-top: 20px;
}

.disabled {
  background-color: var(--disabled-color);
  border-color: var(--disabled-color);
  pointer-events: none;
  user-select: none;
}

.valid {
  box-shadow: 0 0 7px var(--secondary-color);
}

.invalid {
  box-shadow: 0 0 7px var(--error-color);
}

.formInput {
  display: flex;
  flex-flow: row nowrap;
  gap: 20px;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin-bottom: 15px;
}

.input {
  width: 100%;
  height: 30px;
  padding: 0 5px;
  font-family: var(--primary-font);
  font-size: 15px;
  text-align: end;
  color: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 5px;
}

.input.noMode {
  border: 1px solid var(--background-color);
  box-shadow: none;
}

.input.noMode:focus {
  outline: none;
}

.newInput {
  border: none;
}

.errorMsg {
  width: 100%;
  margin: -20px 0 -10px;
  font-size: 14px;
  text-align: right;
  color: var(--error-color);
}

.invalid,
.input.invalid:focus {
  box-shadow: 0 0 7px var(--error-color);
}

.buttonBlock {
  display: flex;
  flex-flow: row nowrap;
  gap: 15px;
  justify-content: center;
  align-items: center;
  width: 100%;
}

@media (width <= 900px) {
  .container {
    min-width: 330px;
  }

  .form,
  .profileBlock {
    flex-flow: column nowrap;
  }

  .chooseInfoBlock {
    flex-flow: row wrap;
    width: 100%;
  }

  .detailInfoBlock {
    width: 100%;
  }

  .chooseInfoItem {
    flex: 1 1 45%;
    box-sizing: border-box;
    max-width: 250px;
    height: 80px;
  }
}

h2 {
  margin: 0;
}
