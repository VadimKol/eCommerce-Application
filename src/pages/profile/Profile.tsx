import type {
  Address,
  ClientResponse,
  Customer,
  MyCustomerUpdate,
  MyCustomerUpdateAction,
} from '@commercetools/platform-sdk';
import { zodResolver } from '@hookform/resolvers/zod';
import classNames from 'classnames';
import { useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { toast } from 'react-toastify';

import { crudAddress, profile } from '@/api/profile.ts';

import { FormChangePassword } from '../../components/form-change-password/FormChangePassword.tsx';
import { FormProfileAddresses } from '../../components/form-profile-addresses/FormProfileAddresses.tsx';
import type { RegisterSchema } from './register-schema.ts';
import { registerSchema } from './register-schema.ts';
import styles from './styles.module.scss';
import type { AddressCustom, CustomerProfile } from './types.ts';

export function Profile(): JSX.Element {
  const [personStatus, setPersonStatus] = useState(true);
  const [shippingStatus, setShippingStatus] = useState(false);
  const [billingStatus, setBillingStatus] = useState(false);
  const [passwordStatus, setPasswordStatus] = useState(false);

  const handleBlock = (person: boolean, shipping: boolean, billing: boolean, password: boolean): void => {
    setPersonStatus(person);
    setShippingStatus(shipping);
    setBillingStatus(billing);
    setPasswordStatus(password);
  };

  const [personInfo, setPersonInfo] = useState<CustomerProfile>({
    version: 1,
    firstName: '',
    lastName: '',
    dateOfBirth: '',
    email: '',
    password: '',
    defaultShippingAddressId: 'you may choose',
    defaultBillingAddressId: 'you may choose',
    billingAddressIds: [],
    shippingAddressIds: [],
  });

  const [addressesShip, setAddressesShip] = useState<AddressCustom[]>([]);
  const [addressesBill, setAddressesBill] = useState<AddressCustom[]>([]);
  const [modeFix, setModeFix] = useState(false);

  const {
    register,
    getFieldState,
    watch,
    trigger,
    setValue,
    formState: { errors, isValid },
  } = useForm<RegisterSchema>({ mode: 'onChange', resolver: zodResolver(registerSchema) });
  const { onChange: onChangeEmail, name: Email, ref: refEmail } = register('email');
  const { onChange: onChangeName, name: Name, ref: refName } = register('name');
  const { onChange: onChangeSurname, name: Surname, ref: refSurname } = register('surname');
  const { onChange: onChangeData, name: Data, ref: refData } = register('age');

  const emailValue = getFieldState('email');
  const nameValue = getFieldState('name');
  const surnameValue = getFieldState('surname');
  const ageValue = getFieldState('age');

  let emailClass = styles.input;
  let nameClass = styles.input;
  let surnameClass = styles.input;
  let ageClass = styles.input;

  if (emailValue.isDirty) {
    emailClass += emailValue.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  if (nameValue.isDirty) {
    nameClass += nameValue.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  if (surnameValue.isDirty) {
    surnameClass += surnameValue.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  if (ageValue.isDirty) {
    ageClass += ageValue.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  useEffect(() => {
    const findAddress = (addresses: Address[], id: string): AddressCustom | undefined =>
      (addresses as AddressCustom[]).find((address) => address.id === id);

    const findAllAddressForOptions = (addresses: Address[], addressIds: string[]): AddressCustom[] => {
      const addressOptions: AddressCustom[] = [];
      addressIds.forEach((addressId: string) => {
        const address = findAddress(addresses, addressId);
        if (address) {
          addressOptions.push(address);
        }
      });
      return addressOptions;
    };
    const fetchProfile = async (): Promise<void> => {
      try {
        const response: ClientResponse<Customer> = await profile();
        const customer: Customer = response?.body;

        if (customer) {
          setPersonInfo({
            version: customer.version || 1,
            firstName: customer.firstName || '',
            lastName: customer.lastName || '',
            dateOfBirth: customer.dateOfBirth || '',
            email: customer.email || '',
            password: customer.password || '',
            defaultShippingAddressId: customer.defaultShippingAddressId || '',
            defaultBillingAddressId: customer.defaultBillingAddressId || '',
            billingAddressIds: customer.billingAddressIds || [],
            shippingAddressIds: customer.shippingAddressIds || [],
          });
          setValue('name', customer.firstName || '');
          setValue('surname', customer.lastName || '');
          setValue('age', customer.dateOfBirth || '');
          setValue('email', customer.email || '');

          const addressOptionsBill = findAllAddressForOptions(customer.addresses, customer.billingAddressIds || []);
          const addressOptionsShip = findAllAddressForOptions(customer.addresses, customer.shippingAddressIds || []);

          setAddressesShip(addressOptionsShip);
          setAddressesBill(addressOptionsBill);
        } else {
          throw new Error('Customer data not available');
        }
      } catch (error) {
        toast('Profile error', { type: 'error' });
      }
    };

    fetchProfile().catch(() => {});
  }, [setValue]);

  const handleChangeInfo = (): void => {
    const formValues = watch();
    const actions: MyCustomerUpdateAction[] = [
      {
        action: 'changeEmail',
        email: formValues.email,
      },
      {
        action: 'setFirstName',
        firstName: formValues.name,
      },
      {
        action: 'setLastName',
        lastName: formValues.surname,
      },
      {
        action: 'setDateOfBirth',
        dateOfBirth: formValues.age,
      },
    ];

    const setDefaultBody: MyCustomerUpdate = {
      version: personInfo.version,
      actions,
    };

    crudAddress(setDefaultBody)
      .then((response: ClientResponse<Customer>) => {
        if (response) {
          toast('User information successfully changed', { type: 'success' });
          setPersonInfo({
            version: response.body.version || 1,
            firstName: response.body.firstName || '',
            lastName: response.body.lastName || '',
            dateOfBirth: response.body.dateOfBirth || '',
            email: response.body.email || '',
            password: response.body.password || '',
            defaultShippingAddressId: response.body.defaultShippingAddressId || '',
            defaultBillingAddressId: response.body.defaultBillingAddressId || '',
            billingAddressIds: response.body.billingAddressIds || [],
            shippingAddressIds: response.body.shippingAddressIds || [],
          });
          setValue('name', response.body.firstName || '');
          setValue('surname', response.body.lastName || '');
          setValue('age', response.body.dateOfBirth || '');
          setValue('email', response.body.email || '');
        }
      })
      .catch((err: Error) => {
        toast(`Error changing information: ${err.message}`, { type: 'error' });
      });
  };
  const handleMode = async (mode: boolean): Promise<void> => {
    setModeFix(mode);
    await trigger();
  };

  const handleCancel = async (): Promise<void> => {
    setValue('name', personInfo.firstName || '');
    setValue('surname', personInfo.lastName || '');
    setValue('age', personInfo.dateOfBirth || '');
    setValue('email', personInfo.email || '');
    await handleMode(!modeFix);
  };

  return (
    <main className={classNames('main', styles.main)}>
      <div className={styles.container}>
        <h1>Profile</h1>
        <div className={styles.profileBlock}>
          <div className={styles.chooseInfoBlock}>
            <button
              type="button"
              onClick={() => handleBlock(true, false, false, false)}
              className={personStatus ? `${styles.chooseInfoItem} ${styles.activeBtn}` : styles.chooseInfoItem}
            >
              Person
            </button>
            <button
              type="button"
              onClick={() => handleBlock(false, true, false, false)}
              className={shippingStatus ? `${styles.chooseInfoItem} ${styles.activeBtn}` : styles.chooseInfoItem}
            >
              Shipping
            </button>
            <button
              type="button"
              onClick={() => handleBlock(false, false, true, false)}
              className={billingStatus ? `${styles.chooseInfoItem} ${styles.activeBtn}` : styles.chooseInfoItem}
            >
              Billing
            </button>
            <button
              type="button"
              onClick={() => handleBlock(false, false, false, true)}
              className={passwordStatus ? `${styles.chooseInfoItem} ${styles.activeBtn}` : styles.chooseInfoItem}
            >
              Password
            </button>
          </div>
          <div className={styles.detailInfoBlock}>
            {personStatus && (
              <form
                onSubmit={(event) => {
                  event.preventDefault();
                  if (isValid) {
                    handleChangeInfo();
                  } else {
                    toast('Invalid form', { type: 'error' });
                  }
                }}
                className={styles.form}
              >
                <div className={styles.detailPerson}>
                  <div className={styles.mainFormProfile}>
                    <div className={classNames(styles.inputWithError, styles.bigInput)}>
                      <label htmlFor="name" className={styles.formInput}>
                        <div className={styles.requiredTitle}>Name:</div>
                        <input
                          id="name"
                          className={modeFix ? nameClass : `${nameClass} ${styles.noMode}`}
                          onChange={(event) => {
                            onChangeName(event).catch(() => {});
                          }}
                          ref={refName}
                          name={Name}
                          type="text"
                          placeholder="John"
                          aria-invalid={errors.name || !nameValue.isDirty ? 'true' : 'false'}
                          required
                          readOnly={!modeFix}
                        />
                      </label>
                      {errors.name && (
                        <span role="alert" className={styles.errorMsg}>
                          {errors.name.message}
                        </span>
                      )}
                    </div>
                    <div className={`${styles.inputWithError}  ${styles.bigInput}`}>
                      <label htmlFor="last" className={styles.formInput}>
                        <div className={styles.requiredTitle}>Surname:</div>
                        <input
                          id="last"
                          onChange={(event) => {
                            onChangeSurname(event).catch(() => {});
                          }}
                          ref={refSurname}
                          name={Surname}
                          type="text"
                          className={modeFix ? surnameClass : `${surnameClass} ${styles.noMode}`}
                          aria-invalid={errors.surname || !surnameValue.isDirty ? 'true' : 'false'}
                          placeholder="Smith"
                          required
                          readOnly={!modeFix}
                        />
                      </label>
                      {errors.surname && (
                        <span role="alert" className={styles.errorMsg}>
                          {errors.surname.message}
                        </span>
                      )}
                    </div>

                    <div className={`${styles.inputWithError}  ${styles.smallInput}`}>
                      <label htmlFor="date" className={styles.formInput}>
                        <div className={styles.requiredTitle}>Birthday:</div>
                        <input
                          id="date"
                          onChange={(event) => {
                            onChangeData(event).catch(() => {});
                          }}
                          ref={refData}
                          name={Data}
                          type={modeFix ? 'date' : 'text'}
                          className={modeFix ? ageClass : classNames(ageClass, styles.noMode)}
                          aria-invalid={errors.age || !ageValue.isDirty ? 'true' : 'false'}
                          required
                          readOnly={!modeFix}
                        />
                      </label>
                      {errors.age && (
                        <span role="alert" className={styles.errorMsg}>
                          {errors.age.message}
                        </span>
                      )}
                    </div>

                    <div className={`${styles.inputWithError}  ${styles.bigInput}`}>
                      <label htmlFor="mail" className={styles.formInput}>
                        <div className={styles.requiredTitle}>Email:</div>
                        <input
                          onChange={(event) => {
                            onChangeEmail(event).catch(() => {});
                          }}
                          id="mail"
                          ref={refEmail}
                          type="text"
                          name={Email}
                          className={modeFix ? emailClass : classNames(emailClass, styles.noMode)}
                          autoComplete="email"
                          placeholder="user@example.com"
                          aria-invalid={errors.email || !emailValue.isDirty ? 'true' : 'false'}
                          required
                          readOnly={!modeFix}
                        />
                      </label>
                      {errors.email && (
                        <span role="alert" className={styles.errorMsg}>
                          {errors.email.message}
                        </span>
                      )}
                    </div>
                    <div className={styles.buttonBlock}>
                      <button
                        onClick={() => {
                          handleMode(!modeFix).catch(() => {});
                        }}
                        type={modeFix ? 'button' : 'submit'}
                        className={!isValid && modeFix ? `${styles.button} ${styles.disabled}` : styles.button}
                      >
                        {modeFix ? 'Save' : 'Change'}
                      </button>
                      {modeFix && (
                        <button
                          onClick={() => {
                            handleCancel().catch(() => {});
                          }}
                          type="button"
                          className={styles.button}
                        >
                          Cancel
                        </button>
                      )}
                    </div>
                  </div>
                </div>
              </form>
            )}
            {shippingStatus && (
              <FormProfileAddresses
                version={personInfo.version}
                addresses={addressesShip}
                defaultAddress={personInfo.defaultShippingAddressId}
                isBilling={false}
                setPersonInfo={setPersonInfo}
                setAddressesShip={setAddressesShip}
                setAddressesBill={setAddressesBill}
              />
            )}
            {billingStatus && (
              <FormProfileAddresses
                version={personInfo.version}
                addresses={addressesBill}
                defaultAddress={personInfo.defaultBillingAddressId}
                isBilling
                setPersonInfo={setPersonInfo}
                setAddressesShip={setAddressesShip}
                setAddressesBill={setAddressesBill}
              />
            )}
            {passwordStatus && (
              <FormChangePassword email={personInfo.email} version={personInfo.version} setPersonInfo={setPersonInfo} />
            )}
          </div>
        </div>
      </div>
    </main>
  );
}
