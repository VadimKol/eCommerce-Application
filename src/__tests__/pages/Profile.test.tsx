import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { Profile } from '@/pages/profile/Profile';

it('renders correctly', () => {
  const { container } = render(
    <BrowserRouter>
      <Profile />
    </BrowserRouter>,
  );
  expect(container).toMatchSnapshot();
});
