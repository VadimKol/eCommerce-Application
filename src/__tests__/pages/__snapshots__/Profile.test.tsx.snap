// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <main
    class="main main"
  >
    <div
      class="container"
    >
      <h1>
        Profile
      </h1>
      <div
        class="profileBlock"
      >
        <div
          class="chooseInfoBlock"
        >
          <button
            class="chooseInfoItem activeBtn"
            type="button"
          >
            Person
          </button>
          <button
            class="chooseInfoItem"
            type="button"
          >
            Shipping
          </button>
          <button
            class="chooseInfoItem"
            type="button"
          >
            Billing
          </button>
          <button
            class="chooseInfoItem"
            type="button"
          >
            Password
          </button>
        </div>
        <div
          class="detailInfoBlock"
        >
          <form
            class="form"
          >
            <div
              class="detailPerson"
            >
              <div
                class="mainFormProfile"
              >
                <div
                  class="inputWithError bigInput"
                >
                  <label
                    class="formInput"
                    for="name"
                  >
                    <div
                      class="requiredTitle"
                    >
                      Name:
                    </div>
                    <input
                      aria-invalid="true"
                      class="input noMode"
                      id="name"
                      name="name"
                      placeholder="John"
                      readonly=""
                      required=""
                      type="text"
                    />
                  </label>
                </div>
                <div
                  class="inputWithError  bigInput"
                >
                  <label
                    class="formInput"
                    for="last"
                  >
                    <div
                      class="requiredTitle"
                    >
                      Surname:
                    </div>
                    <input
                      aria-invalid="true"
                      class="input noMode"
                      id="last"
                      name="surname"
                      placeholder="Smith"
                      readonly=""
                      required=""
                      type="text"
                    />
                  </label>
                </div>
                <div
                  class="inputWithError  smallInput"
                >
                  <label
                    class="formInput"
                    for="date"
                  >
                    <div
                      class="requiredTitle"
                    >
                      Birthday:
                    </div>
                    <input
                      aria-invalid="true"
                      class="input noMode"
                      id="date"
                      name="age"
                      readonly=""
                      required=""
                      type="text"
                    />
                  </label>
                </div>
                <div
                  class="inputWithError  bigInput"
                >
                  <label
                    class="formInput"
                    for="mail"
                  >
                    <div
                      class="requiredTitle"
                    >
                      Email:
                    </div>
                    <input
                      aria-invalid="true"
                      autocomplete="email"
                      class="input noMode"
                      id="mail"
                      name="email"
                      placeholder="user@example.com"
                      readonly=""
                      required=""
                      type="text"
                    />
                  </label>
                </div>
                <div
                  class="buttonBlock"
                >
                  <button
                    class="button"
                    type="submit"
                  >
                    Change
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
</div>
`;
